File: default/server/_jscript/Site.js
(function() {
  var mongodb, node_url, smio;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };
  require('./shared/PromiseProxy');
  require('./Database');
  mongodb = require('mongodb');
  node_url = require('url');
  smio = global.smoothio;
  smio.Site = (function() {
    function Site(session, url, rc) {
      this.session = session;
      this.url = url;
      this.rc = rc;
      this.server = this.session.server;
      this.doc = null;
      this.inst = this.server.inst;
      this.dbAdmin = this.inst.mongos['admin'];
      this.dbServer = this.inst.mongos["smoothio__" + this.server.serverName];
      this.dbShared = this.inst.mongos['smoothio_shared'];
      if (!this.url) {
        this.url = '/';
      }
      this.url = this.url.toLowerCase();
      this.uri = node_url.parse(this.url);
    }
    Site.prototype.checkExists = function(cb_err_hasSites) {
      if (this.doc) {
        return cb_err_hasSites(null, true);
      } else {
        return this.dbServer.withCollection("_smio_sites", __bind(function(err, col) {
          var makeQuery;
          if (err) {
            return cb_err_hasSites(err);
          }
          makeQuery = function(url) {
            return function() {
              return (url.indexOf(this._smio_url)) === 0;
            };
          };
          return (col.find({
            $where: makeQuery(this.url)
          })).toArray(__bind(function(err, results) {
            if (err) {
              return cb_err_hasSites(err);
            }
            if (results && results.length) {
              this.doc = (_.sortBy(results, function(doc) {
                return -doc._smio_url.length;
              }))[0];
              return cb_err_hasSites(null, true);
            } else {
              return col.find().nextObject(__bind(function(err, doc) {
                return cb_err_hasSites(err, doc ? true : false);
              }, this));
            }
          }, this));
        }, this));
      }
    };
    Site.prototype.getControlUpdates = function(sinceTicks, cb) {
      var ct;
      if (sinceTicks) {
        return cb(null, {});
      } else {
        ct = "SmoothioCore_CommonControls_mainframe";
        return this.checkExists(function(err, serverHasSites) {
          if (!serverHasSites) {
            ct = "SmoothioCore_ServerSetup_initialserversetup";
          }
          return cb(err, {
            "": {
              "_": ct
            }
          });
        });
      }
    };
    return Site;
  })();
}).call(this);

File: _cscript/Site.cs

require './shared/PromiseProxy'
require './Database'
mongodb = require 'mongodb'
node_url = require 'url'

smio = global.smoothio

class smio.Site
	constructor: (@session, @url, @rc) ->
		@server = @session.server
		@doc = null
		@inst = @server.inst
		@mongo = @inst.getDbServer()
		@dbAdmin = @inst.getDb @mongo, 'admin'
		@dbServer = @inst.getDb @mongo, "smoothio__#{@server.serverName}"
		@dbShared = @inst.getDb @mongo, 'smoothio_shared'
		if not @url
			@url = '/'
		@url = @url.toLowerCase()
		@uri = node_url.parse @url

	checkExists: (cb_err_hasSites) ->
		if @doc
			cb_err_hasSites null, true
		else
			@dbServer.withCollection "_smio_sites", (err, col) =>
				return cb_err_hasSites err if err
				makeQuery = (url) -> -> (url.indexOf this._smio_url) is 0
				(col.find $where: (makeQuery @url)).toArray (err, results) =>
					return cb_err_hasSites err if err
					if results and results.length
						@doc = (_.sortBy results, (doc) -> -doc._smio_url.length)[0]
						cb_err_hasSites null, true
					else
						col.find().nextObject (err, doc) =>
							cb_err_hasSites err, if doc then true else false	

	getControlUpdates: (sinceTicks, cb) ->
		if sinceTicks
			cb null, {}
		else
			ct = "Core_Controls_MainFrame"
			@checkExists (err, serverHasSites) ->
				if not serverHasSites
					ct = "Core_ServerSetup_InitialSiteSetup"
				cb err, { "": { "_": ct } }

